homeworks/Assignment4_Final.R

```diff
@@ -0,0 +1,162 @@
+###############################################################################################
+# Assignment 4
+# November 11th, 2020
+###############################################################################################
+rm(list = ls())
+###############################################################################################
+# Loading Libraries
+# install.packages as necessary(Updated R version)
+install.packages("dplyr")
+install.packages("tidyverse")
+install.packages("data.table")
+install.packages("broom")
+install.packages("lintr")
+install.packages("reshape")
+###############################################################################################
+library(lintr)
+library(broom)
+library(tidyverse)
+library(stats)
+library(dplyr)  
+library(data.table)
+library(tidyr)
+library(reshape)
+library(ggplot2)
+###############################################################################################
+#Problem 1
+###############################################################################################
+# data - data frame that contains all the information
+# group_id - A vector with one or multiple column names that uniquely define the groups.
+# obs - A vector with one or multiple column names of targeting observations
+# fun - A summary function
+# ... - Extra arguments for the summary function
+## The function should return a data frame with 'group_id' column(s) and all the summary staistics
+summarise_group_data <- function(data, group_id, obs, fun, ...){
+  data %>%
+    group_by_at(group_id) %>%
+    summarise_at(.vars = vars(obs), .funs = fun, ...)
+}
+
+#Testing
+data(iris)
+summarise_group_data(data = iris, group_id = 'Species', obs = c('Petal.Length', 'Sepal.Length'), fun = sd)
+
+###############################################################################################
+#Problem 2
+###############################################################################################
+regress_group_data <- function(data, group_id, obs, covs, include_intercept = TRUE, ...){
+  formula <- paste0(obs, '~', covs)
+  covs <- paste0(covs, collapse = '+')
+  final <- data %>% group_by_at(group_id) %>%
+    do(tidy(lm(formula = as.formula(formula), data = . ))) %>%
+    select(-std.error, -statistic, -p.value)
+  mod <- pivot_wider(final, names_from = 'term', values_from = 'estimate')
+  return(mod)
+  if(!include_intercept)
+    final_ni <- select(.data = mod, -intercept)
```

Comment:
`include_intercept` should happen in the fitting step instead of extraction of result step. -0.1